### PR TITLE
OS/NestedContact: Cleanup

### PR DESCRIPTION
**Pull request by drdanz**
* Use PIMPL
* Move implementation in .cpp
* Cleanup

### CODE REVIEW

File: src/libYARP_OS/src/NestedContact.cpp [hunk 1]
```diff
@@ -53,33 +62,137 @@ bool NestedContact::fromString(const ConstString& nFullName) {
...
+    fromString(fullName);
+}
+
+NestedContact::NestedContact(const NestedContact& rhs) :
+        mPriv(new Private(*(rhs.mPriv)))
+{
+}
+
+NestedContact::NestedContact(NestedContact&& rhs) :
+        mPriv(new Private(std::move(*(rhs.mPriv))))
```
Comment (traversaro) [inline src/libYARP_OS/src/NestedContact.cpp]: I am a bit confused by this.  The move constructor should have the advantage to avoid the `new`?

Comment (drdanz) [reply src/libYARP_OS/src/NestedContact.cpp]: good catch...
Probably it should just be

```c++
NestedContact::NestedContact(NestedContact&& rhs) :
        mPriv(std::move(rhs.mPriv))
```

Comment (drdanz) [reply src/libYARP_OS/src/NestedContact.cpp]: Probably also `NestedContact::NestedContact(NestedContact&& rhs) = default` should work...

Comment (traversaro) [reply src/libYARP_OS/src/NestedContact.cpp]: I don't think `mPriv(std::move(rhs.mPriv))` will work, see https://stackoverflow.com/questions/9452503/why-does-moving-a-pointer-variable-not-set-it-to-null . I guess the default version should work, but I am not really sure about how that will work.

Comment (drdanz) [reply src/libYARP_OS/src/NestedContact.cpp]: @traversaro now I remember the problem...
The problem was that the values are swapped, therefore the rvalue pointer is replaced with an uninitialized pointer, then it is deleted, hence crash.
I replaced it with the `new Private`, but of course this does not make sense for the move constructor.
I think this should work:
```c++
NestedContact::NestedContact(NestedContact&& rhs) :
        mPriv(std::move(rhs.mPriv))
{
    rhs.mPriv = nullptr;
}
```
@traversaro does it look ok to you?

Comment (drdanz) [reply src/libYARP_OS/src/NestedContact.cpp]: BTW, at some point I'd like to come to this... https://oliora.github.io/2015/12/29/pimpl-and-rule-of-zero.html

**Pull request merged**